File: client_jms/test/TopicListenerTest.java
/*
 * Copyright (c) 2013-2021 Contributors to the Eclipse Foundation
 * 
 * See the NOTICE file(s) distributed with this work for additional
 * information regarding copyright ownership.
 * 
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0
 * 
 * SPDX-License-Identifier: EPL-2.0
 */


import javax.jms.BytesMessage;
import javax.jms.Connection;
import javax.jms.ConnectionFactory;
import javax.jms.DeliveryMode;
import javax.jms.ExceptionListener;
import javax.jms.JMSException;
import javax.jms.Message;
import javax.jms.MessageConsumer;
import javax.jms.MessageListener;
import javax.jms.MessageProducer;
import javax.jms.Session;
import javax.jms.Topic;

import com.ibm.ima.jms.ImaJmsFactory;
import com.ibm.ima.jms.ImaProperties;


/*
 * Test a point-to-point send and receive with acknowledge
 * 
 * There are three positional arguments which can be set when running this test:
 * 1. The size of the message (default = 128)
 * 2. The number of messages (default = 100)
 * 3. The name of the topic (default = "RequestT")
 * 
 * The location of the IMA server can be specified using two environment variables:
 * IMAServer - The IP address of the server (default = 127.0.0.1)
 * IMAPort   - The port number of the server (default = 16102)
 * 
 * The default can also be modified in the section below.
 */
public class TopicListenerTest implements ExceptionListener {
	static int  msglen = 128;    
	static int  count = 100;
	static String topicName = "RequestT";

	static ConnectionFactory   fact;
	static Connection          conn;
	static volatile boolean recvdone1;
	static volatile boolean recvdone2;
	
	static int  sendcount;
	static int  recvcount1;
	static int  recvcount2;
	
    /*
     * Send messages
     */
    public static long doSend(Session sess, MessageProducer prod) throws JMSException {
    	prod.setDeliveryMode(DeliveryMode.NON_PERSISTENT);
    	conn.start();
    	//conn2.start();
    	try { Thread.sleep(100); } catch (Exception e) {}
    	
    	long startTime = System.currentTimeMillis();
    	
    	BytesMessage bmsg = sess.createBytesMessage();
    	byte [] msgbytes = new byte[msglen];
    	
    	for (int i=0; i<count; i++) {
    		bmsg.clearBody();
    		msgbytes[0] = (byte)i;
    		bmsg.writeBytes(msgbytes);
    		prod.send(bmsg);
    		sendcount++;
    		if (sendcount%500 == 0)
    			try { Thread.sleep(3);} catch (Exception e) {} 
    	}
    	
    	int wait = 800;
    	while (!(recvdone1 && recvdone2) && wait-- > 0) {
    		long sleepTime = 50;
    		long curTime = System.currentTimeMillis();
    		long wakeUpTime = curTime + sleepTime;

    		while (wakeUpTime > curTime) {
    			try { Thread.sleep(wakeUpTime - curTime); } catch (Exception e) {}
    			curTime = System.currentTimeMillis();
    		}
    	}
    	long endTime = System.currentTimeMillis();
    	if (!recvdone1) {
    		System.out.println("Recv(1) - Messages were not all received");
    	}
    	if (!recvdone2) {
            System.out.println("Recv(2) - Messages were not all received");
    	}
        return endTime-startTime;  	
    }
    
    /*
     * Parse arguments
     */
    public static void parseArgs(String [] args) {
     	if (args.length > 0) {
    		try { 
    			msglen = Integer.valueOf(args[0]);
    		} catch (Exception e) {
    			syntaxhelp("Message length not valid");
    			return;
    		}
        }
    	if (args.length > 1) {
    		try { 
    			count = Integer.valueOf(args[1]);
    		} catch (Exception e) {
    			syntaxhelp("");
    			return;
    		}
        }
    	if (args.length > 2) {
    		topicName = args[2];
    	}
    }
    
    /*
     * Put out a simple syntax help
     */
    public static void syntaxhelp(String msg)  {
    	if (msg != null)
            System.out.println(msg);
    	System.out.println("PtPListenerTest  msglen  count ");
    }
    
    /*
     * Main method 
     */
    public static void main(String [] args) {
    	long millis = 0;
    	Session sendSess = null;
    	Session recvSess1 = null;
    	Session recvSess2 = null;
    	MessageProducer prod = null;
    	MessageConsumer cons1 = null;
    	MessageConsumer cons2 = null;
    	Topic prodT;
    	Topic consT1;
    	Topic consT2;
    	
    	parseArgs(args);
    	/*
    	 * Create connection factory and connection
    	 */
    	try {
    	    fact = ImaJmsFactory.createConnectionFactory();
    	    String server = System.getenv("IMAServer");
    	    String port   = System.getenv("IMAPort");
    	    String transport = System.getenv("IMATransport");
    	    String recvBuffer = System.getenv("IMARecvBuffer");
    	    String sendBuffer = System.getenv("IMASendBuffer");
            if (server == null)
                server = "127.0.0.1";
            if (port == null)
                port = "16102";
            ((ImaProperties)fact).put("Server", server);
            ((ImaProperties)fact).put("Port", port);
    	    if (transport != null)
    	    	((ImaProperties)fact).put("Protocol", transport);
    	    if (recvBuffer != null)
    	    	((ImaProperties)fact).put("RecvBuffer", recvBuffer);
    	    if (sendBuffer != null)
    	    	((ImaProperties)fact).put("SendBuffer", sendBuffer);

    	    ((ImaProperties)fact).put("DisableMessageTimestamp", true);
            ((ImaProperties)fact).put("DisableMessageID", true);
    	    conn = fact.createConnection();
        	sendSess = conn.createSession(false, Session.AUTO_ACKNOWLEDGE);
        	recvSess1 = conn.createSession(false, Session.AUTO_ACKNOWLEDGE);
        	recvSess2 = conn.createSession(false, Session.AUTO_ACKNOWLEDGE);
        	
        	prodT = sendSess.createTopic(topicName); 
        	consT1 = recvSess1.createTopic(topicName);
        	((ImaProperties)consT1).put("DisableACK", true);
        	consT2 = recvSess2.createTopic(topicName);
//        	((ImaProperties)consT2).put("DisableACK", true);
        	prod = sendSess.createProducer(prodT);
        	
        	conn.setExceptionListener(new TopicListenerTest());
        	
        	cons1 = recvSess1.createConsumer(consT1);
        	MessageListener msgListener1 = new TopicListenerTest().new TopicMessageListener();
        	((TopicMessageListener)msgListener1).setInstance(1);
			cons1.setMessageListener(msgListener1);
			
			cons2 = recvSess2.createConsumer(consT2);
			MessageListener msgListener2 = new TopicListenerTest().new TopicMessageListener();
            ((TopicMessageListener)msgListener2).setInstance(2);
            cons2.setMessageListener(msgListener2);
    	} catch (JMSException jmse) {
    		jmse.printStackTrace(System.out);
    		return;
    	}
    	
    	/*
    	 * Run the send
    	 */
    	try {
    	    millis = doSend(sendSess, prod);
    	} catch (Exception ex) {
    		ex.printStackTrace(System.out);
    	}
    	
    	/*
    	 * Print out statistics
    	 */
    	System.out.println("Time = "+millis);
    	System.out.println("Send = "+sendcount);
    	System.out.println("Recv(1) = "+recvcount1);
    	System.out.println("Recv(2) = "+recvcount2);
    	try {
    	    conn.close();
    	} catch (Exception e) {
    		e.printStackTrace();
    	}
    }

    class TopicMessageListener implements MessageListener {
        int recvcount;
        int inst;
        boolean done = false;
        
        void setInstance(int inst) {
            this.inst = inst;
        }

        public void onMessage(Message msg) {
            if(inst == 1) {
                recvcount = recvcount1;
                done = recvdone1;
            } else {
                recvcount = recvcount2;
                done = recvdone2;
            }
            /* Validate the message, this will throw ClassCastException if it is not a BytesMessage */
            BytesMessage bmsg = (BytesMessage)msg;
            byte [] b = new byte [1];
            try {
                bmsg.readBytes(b);
                if (bmsg.getBodyLength() != msglen || b[0] != (byte)recvcount) {
                    throw new JMSException("Message received is not the correct message: expected("+recvcount+") found("+b[0]+") expected length("+msglen+") found length("+bmsg.getBodyLength()+")");
                }
            }
            catch (JMSException jmse) {
                throw new RuntimeException(jmse);
            }
            recvcount++;
            if (recvcount == count) {
                done = true;
            }
            if(inst == 1) {
                recvcount1 = recvcount;
                recvdone1 = done;
            } else {
                recvcount2 = recvcount;
                recvdone2 = done;
            }
        }
    }

	@Override
	public void onException(JMSException arg0) {

		System.out.println("Exception - " + arg0.getMessage());
	}
}

File: server_ismc/test/topiclistener_test.c
/*
 * Copyright (c) 2012-2021 Contributors to the Eclipse Foundation
 * 
 * See the NOTICE file(s) distributed with this work for additional
 * information regarding copyright ownership.
 * 
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0
 * 
 * SPDX-License-Identifier: EPL-2.0
 */

#include <ismc.h>
#include <ismc_p.h>

#include <ismutil.h>
#include <transport.h>
/*
 * Test a point-to-point send and receive with acknowledge
 *
 * There are three positional arguments which can be set when running this test:
 * 1. The size of the message (default = 128)
 * 2. The number of messages (default = 100)
 * 3. The name of the topic (default = "RequestT")
 *
 * The location of the ISM server can be specified using two environment variables:
 * ISMServer - The IP address of the server (default = 127.0.0.1)
 * ISMPort   - The port number of the server (default = 16102)
 *
 * The default can also be modified in the section below.
 */

static int  msglen = 128;
static int  count = 100;
static const char * topicName = "RequestT";

static ismc_connection_t * conn;
static volatile int recvdone1 = 0;
static volatile int recvdone2 = 0;

static int  sendcount = 0;
static int  recvcount1 = 0;
static int  recvcount2 = 0;

#define CLIENT_ID			"TopicListenerTest"
#define SERVER_ADDRESS		"127.0.0.1"
#define DEFAULT_PORT		16102

int parseArgs(int argc, char * * argv);
void syntaxhelp(const char * msg);
uint64_t doSend(ismc_session_t * sess, ismc_producer_t * prod);
void messageListener(ismc_message_t * message, ismc_consumer_t * consumer, void * userdata);
void ism_engine_threadInit(uint8_t isStoreCrit)
{
}
int ism_process_monitoring_action(ism_transport_t *transport, const char *inpbuf, int buflen, concat_alloc_t *output_buffer, int *rc)
{
	return 0;
}
/*
 * Main method
 */
int main(int argc, char * * argv) {
	uint64_t millis = 0;
	ismc_session_t *  sendSess = NULL;
	ismc_session_t *  recvSess1 = NULL;
	ismc_session_t *  recvSess2 = NULL;
	ismc_producer_t * prod = NULL;
	ismc_consumer_t * cons1 = NULL;
	ismc_consumer_t * cons2 = NULL;
	ismc_destination_t * prodT;
	ismc_destination_t * consT1;
	ismc_destination_t * consT2;
	char * server;
	char * port;
	char * transport;
	char * recvBuffer;
	char * sendBuffer;
	int    serverPort;
	ism_field_t field;
	int rc;
	int id1 = 1, id2 = 2;

	if (parseArgs(argc, argv) < 0) {
		return -1;
	}

	/*
	 * Create connection
	 */

	server = getenv("ISMServer");
	port   = getenv("ISMPort");
	transport = getenv("ISMTransport");
	recvBuffer = getenv("ISMRecvBuffer");
	sendBuffer = getenv("ISMSendBuffer");

	if (!server) {
		server = SERVER_ADDRESS;
	}

	if (port != NULL) {
		serverPort = atoi(port);
	} else {
		serverPort = DEFAULT_PORT;
	}

	conn = ismc_createConnectionX(CLIENT_ID, NULL, server, serverPort);
	if (conn == NULL) {
		printf("Cannot create connection\n");
		return -1;
	}

	if (port) {
		ismc_setStringProperty(conn, "Port", port);
	}

	if (transport) {
		ismc_setStringProperty(conn, "Protocol", transport);
	}

	if (recvBuffer) {
		ismc_setStringProperty(conn, "RecvBuffer", recvBuffer);
	}

	if (sendBuffer) {
		ismc_setStringProperty(conn, "SendBuffer", sendBuffer);
	}

	field.type = VT_Boolean;
	field.val.i = 1;
	ismc_setProperty(conn, "DisableMessageTimestamp", &field);
	ismc_setProperty(conn, "DisableMessageID", &field);

	rc = ismc_connect(conn);
	if (rc != 0) {
		printf("Cannot connect\n");
		return -1;
	}

	sendSess  = ismc_createSession(conn, 0, 1);	// Auto-ACK
	recvSess1 = ismc_createSession(conn, 0, 1);	// Auto-ACK
	recvSess2 = ismc_createSession(conn, 0, 1);	// Auto-ACK
	if (!sendSess || !recvSess1 || !recvSess2) {
		printf("Cannot create session\n");
		return -1;
	}

	prodT  = ismc_createTopic(topicName);
	consT1 = ismc_createTopic(topicName);
	consT2 = ismc_createTopic(topicName);
	if (!prodT || !consT1 || !consT2) {
		printf("Cannot create topic\n");
		return -1;
	}

//	ismc_setProperty(consT1, "DisableACK", &field);
//	ismc_setProperty(consT2, "DisableACK", &field);

	prod = ismc_createProducer(sendSess, prodT);
	if (!prod) {
		printf("Cannot create producer\n");
		return -1;
	}

	cons1 = ismc_createConsumer(recvSess1, consT1, NULL, 0);
	cons2 = ismc_createConsumer(recvSess2, consT2, NULL, 0);
	if (!cons1 || !cons2) {
		printf("Cannot create consumer\n");
		return -1;
	}

	ismc_setlistener(cons1, messageListener, &id1);
	ismc_setlistener(cons2, messageListener, &id2);

	/*
	 * Run the send
	 */
	millis = doSend(sendSess, prod);

	/*
	 * Print out statistics
	 */
	printf("Time = %ld\n", millis);
	printf("Send = %d\n", sendcount);
	printf("Recv(1) = %d\n", recvcount1);
	printf("Recv(2) = %d\n", recvcount2);

	ismc_closeConsumer(cons1);
	ismc_closeConnection(conn);
	ismc_free(prod);
	ismc_free(prodT);
	ismc_free(cons1);
	ismc_free(cons2);
	ismc_free(consT1);
	ismc_free(consT2);
	ismc_free(sendSess);
	ismc_free(recvSess1);
	ismc_free(recvSess2);
	ismc_free(conn);

	return 0;
}

/*
 * Parse arguments
 */
int parseArgs(int argc, char * * argv) {
 	if (argc > 1) {
		msglen = atoi(argv[1]);
		if (msglen <= 0) {
			syntaxhelp("Message length not valid");
			return -1;
		}
    }
	if (argc > 2) {
		count = atoi(argv[2]);
		if (count <= 0) {
			syntaxhelp("");
			return -1;
		}
    }
	if (argc > 3) {
		topicName = argv[3];
	}

	return 0;
}

/*
 * Put out a simple syntax help
 */
void syntaxhelp(const char * msg)  {
	if (msg != NULL)
        printf("%s\n", msg);
	printf("TopicListenerTest  msglen  count\n");
}

void messageListener(ismc_message_t * message, ismc_consumer_t * consumer, void * userdata) {
	int inst = *(int *)userdata;
    int recvcount;
    int done = 0;
    unsigned char b;

    if(inst == 1) {
        recvcount = recvcount1;
        done = recvdone1;
    } else {
        recvcount = recvcount2;
        done = recvdone2;
    }

    /* Validate the message */
	ismc_getContent(message, (char*)&b, 0, 1);
	if (ismc_getContentSize(message) != msglen || b != (unsigned char)(recvcount & 0xFF)) {
		printf("Message received in %d is not the correct message: "
			   "expected(%d) found(%d) "
			   "expected length(%d) found length(%d)\n",
			   inst, (unsigned char)(recvcount&0xFF), b, msglen, ismc_getContentSize(message));

	    if(inst == 1) {
	        recvdone1 = 1;
	    } else {
	        recvdone2 = 1;
	    }

		return;
	}

    recvcount++;
    if (recvcount == count) {
        done = 1;
    }
    if(inst == 1) {
        recvcount1 = recvcount;
        recvdone1 = done;
    } else {
        recvcount2 = recvcount;
        recvdone2 = done;
    }
}

/*
 * Send messages
 */
uint64_t doSend(ismc_session_t * sess, ismc_producer_t * prod) {
	int rc;
	int i;
	ism_time_t startTime, endTime;
	ismc_message_t * bmsg;
	char * msgbytes;
	int wait = 800;

	ism_common_sleep(100000);

	rc = ismc_startConnection(conn);
//	if (rc != 0) {
//		printf("Cannot start connection: %d\n", rc);
//		return -1;
//	}

	startTime = ism_common_currentTimeNanos();

	bmsg = ismc_createMessage(sess, MTYPE_BytesMessage);
	ismc_setDeliveryMode(bmsg, ISMC_NON_PERSISTENT);		// Non-persistent
	msgbytes = calloc(msglen, 1);

	for (i = 0; i < count; i++) {
		ismc_clearContent(bmsg);
		msgbytes[0] = (i&0xFF);
		ismc_setContent(bmsg, msgbytes, msglen);

		rc = ismc_send(prod, bmsg);
		if (rc) {
			printf("Send %d failed\n", i);
			break;
		}

		sendcount++;
		if ((sendcount % 500) == 0) {
			ism_common_sleep(3000);
		}
	}

	free(msgbytes);

	while (wait-- > 0 && !recvdone1 && !recvdone2) {
		ism_common_sleep(50000);
	}
	endTime = ism_common_currentTimeNanos();

	ismc_freeMessage(bmsg);

	if (!recvdone1) {
		printf("Recv(1) - Messages were not all received\n");
	}
	if (!recvdone2) {
		printf("Recv(2) - Messages were not all received\n");
	}
	return (endTime - startTime) / 1000000;
}
